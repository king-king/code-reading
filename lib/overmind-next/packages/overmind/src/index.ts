import { IS_DERIVED_CONSTRUCTOR } from './derived'
import * as internalTypes from './internalTypes'
import { Overmind } from './Overmind'
import { IConfiguration } from './types'
import * as utils from './utils'

export * from './types'
export * from './operators'
export { Overmind } from './Overmind'
export {
  ResolveAction,
  OperatorContextFunction,
  ContextFunction,
} from './internalTypes'
export { createOperator, createMutationOperator } from './operator'
export { MODE_DEFAULT, MODE_TEST, MODE_SSR, ENVIRONMENT, json } from './utils'
export { SERIALIZE, rehydrate } from './rehydrate'
export { Statemachine, statemachine } from './statemachine'
export * from './OvermindMock'
export * from './OvermindSSR'

export const derived = <S extends object, R extends object, O>(
  cb: (state: S, rootState: R) => O
): O => {
  cb[IS_DERIVED_CONSTRUCTOR] = true
  return cb as any
}

export function createOvermind<Config extends IConfiguration>(
  config: Config,
  options?: internalTypes.Options
): Overmind<Config> {
  return new Overmind(config, options, { mode: utils.MODE_DEFAULT })
}
