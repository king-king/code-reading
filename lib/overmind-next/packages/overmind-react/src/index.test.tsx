import { Overmind, createOvermindMock, IContext } from 'overmind'
import * as React from 'react'
import * as renderer from 'react-test-renderer'

import { Provider, createStateHook } from './'

describe('React', () => {
  test('should allow using hooks', () => {
    let renderCount = 0

    const doThis = ({ state }: Context) => {
      state.foo = 'bar2'
    }
    const state = {
      foo: 'bar',
    }
    const actions = {
      doThis,
    }
    const config = {
      state,
      actions,
    }

    type Context = IContext<{
      state: typeof state
      actions: typeof actions
    }>

    const app = new Overmind(config)

    const useState = createStateHook<Context>()

    const FooComponent: React.FunctionComponent = () => {
      const state = useState()
      renderCount++

      return <h1>{state.foo}</h1>
    }

    const tree = renderer
      .create(
        <Provider value={app}>
          <FooComponent />
        </Provider>
      )
      .toJSON()

    expect(renderCount).toBe(1)

    renderer.act(() => {
      app.actions.doThis()
    })

    expect(renderCount).toBe(2)
    expect(tree).toMatchSnapshot()
  })
  test('should allow using hooks with scoped tracking', () => {
    let renderCount = 0

    const doThis = ({ state }: Context) => {
      state.foo.push({ foo: 'bar2' })
    }
    const doThat = ({ state }: Context) => {
      state.foo[0].foo = 'bar3'
    }
    const state = {
      foo: [{ foo: 'bar' }],
    }
    const actions = {
      doThis,
      doThat,
    }
    const config = {
      state,
      actions,
    }

    type Context = IContext<{
      state: typeof state
      actions: typeof actions
    }>

    const app = new Overmind(config)

    const useState = createStateHook<Context>()

    const FooComponent: React.FunctionComponent = () => {
      const state = useState((state) => state.foo[0])
      renderCount++

      return <h1>{state.foo}</h1>
    }

    const tree = renderer
      .create(
        <Provider value={app}>
          <FooComponent />
        </Provider>
      )
      .toJSON()

    expect(renderCount).toBe(1)

    renderer.act(() => {
      app.actions.doThis()
    })

    expect(renderCount).toBe(1)

    renderer.act(() => {
      app.actions.doThat()
    })
    expect(renderCount).toBe(2)

    // This is not showing the expected result, but logging the rendering does, so must be the
    // library messing it up
    expect(tree).toMatchSnapshot()
  })
  test('should allow using mocked Overmind', () => {
    let renderCount = 0

    const doThis = ({ state }: Context) => {
      state.foo = 'bar2'
    }
    const state = {
      foo: 'bar',
    }
    const actions = {
      doThis,
    }
    const config = {
      state,
      actions,
    }

    type Context = IContext<{
      state: typeof state
      actions: typeof actions
    }>
    const useState = createStateHook<Context>()

    const FooComponent: React.FunctionComponent = () => {
      const state = useState()
      renderCount++

      return <h1>{state.foo}</h1>
    }

    const mock = createOvermindMock(config)
    const tree = renderer
      .create(
        <Provider value={mock}>
          <FooComponent />
        </Provider>
      )
      .toJSON()

    expect(renderCount).toBe(1)
    expect(tree).toMatchSnapshot()
  })
  test('should throw an error without provider', () => {
    expect.assertions(1)

    const useState = createStateHook()

    const FooComponent = () => {
      const state = useState()

      return <h1>{(state as any).foo}</h1>
    }

    expect(() => {
      renderer.create(<FooComponent />).toJSON()
    }).toThrow(Error)
  })
})
