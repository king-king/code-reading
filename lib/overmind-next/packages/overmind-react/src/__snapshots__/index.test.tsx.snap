// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React should allow using hooks 1`] = `
<h1>
  bar
</h1>
`;

exports[`React should allow using hooks with scoped tracking 1`] = `
<h1>
  bar
</h1>
`;

exports[`React should allow using mocked Overmind 1`] = `
<h1>
  bar
</h1>
`;
